rfc/rfc000-package-server.md

```diff
@@ -1,19 +1,87 @@
+# Package Server
+
 ---
+
 RFC: 000
-Author: Peter Bourgon <peter@bourgon.org>
+
+Author: Matt Heath <matt@mattheath.com>
+
 Status: Draft
----
 
-# package server
+---
 
 ## Motivation
 
-http://peter.bourgon.org/go-kit/#package-server
+[http://peter.bourgon.org/go-kit/#package-server](http://peter.bourgon.org/go-kit/#package-server)
+
+Package server is probably the biggest and most important component of the toolkit. Ideally, we should be able to write our services as implementations of normal, nominal Go interfaces, and delegate integration with the environment to the server package. The package should encode and enforce conventions for server-side concerns, like health checks, system-wide request tracing, connection management, backpressure and throttling, and so on. For each of those topics, it should provide interfaces for different, pluggable strategies. It should integrate with service discovery, and work equally well over multiple transports. Considerable prior art exists in the form of Finagle, Karyon (Netflixs application service library), and likely many more.
 
 ## Scope
 
-To be defined.
+### Endpoints
+
+*   An endpoint is defined as a handler interface which receives a request and returns a response or an error to the client.
+*   A server SHALL have one or more endpoints.
+*   An endpoint SHALL accept a request context and propagate this - allowing access to request-specific values such as the identity of the end user, authorization tokens, and the request's deadline.
+*   An endpoint SHALL respect request cancellation if or when the request's deadline expires.
+*   A server MAY expose information about its endpoints, to allow integration with additional tools.
+
+### Contexts
+
+*   Requests SHALL be executed within a request [context](https://blog.golang.org/context), which the server will pass through the request chain.
+
+### Throttling & Backpressure
+
+*   A server MAY throttle inbound requests and reject requests from clients based on a number of factors.
+*   A server MAY respond with either an Out Of Capacity error, or a Rate Limit Exceeded error when rejecting requests.
+*   A server MAY limit the total number of concurrent requests it can serve.
+*   A server MAY impose rate limits on specific clients.
```

Comment:
the server might not be the right layer for some of these DOS-ish things (i.e., perhaps that belongs at the transport auth layer?)